lib/js-api/src/externals/memory.rs

```diff
@@ -0,0 +1,319 @@
+/// mutable from both host and WebAssembly.
+///
+/// Spec: <https://webassembly.github.io/spec/core/exec/runtime.html#memory-instances>
+#[derive(Debug, Clone, PartialEq)]
+pub struct Memory {
+    store: Store,
+    vm_memory: VMMemory,
+}
+
+impl Memory {
+    /// Creates a new host `Memory` from the provided [`MemoryType`].
+    ///
+    /// This function will construct the `Memory` using the store
+    /// [`BaseTunables`][crate::tunables::BaseTunables].
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # use wasmer_js::{Memory, MemoryType, Pages, Store, Type, Value};
+    /// # let store = Store::default();
+    /// #
+    /// let m = Memory::new(&store, MemoryType::new(1, None, false)).unwrap();
+    /// ```
+    pub fn new(store: &Store, ty: MemoryType) -> Result<Self, MemoryError> {
+        let descriptor = js_sys::Object::new();
+        js_sys::Reflect::set(&descriptor, &"initial".into(), &ty.minimum.0.into()).unwrap();
+        if let Some(max) = ty.maximum {
+            js_sys::Reflect::set(&descriptor, &"maximum".into(), &max.0.into()).unwrap();
+        }
+        js_sys::Reflect::set(&descriptor, &"shared".into(), &ty.shared.into()).unwrap();
+
+        let js_memory = js_sys::WebAssembly::Memory::new(&descriptor)
+            .map_err(|_e| MemoryError::Generic("Error while creating the memory".to_owned()))?;
+
+        let memory = VMMemory::new(js_memory, ty);
+        Ok(Self {
+            store: store.clone(),
+            vm_memory: memory,
+        })
+    }
+
+    /// Returns the [`MemoryType`] of the `Memory`.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # use wasmer_js::{Memory, MemoryType, Pages, Store, Type, Value};
+    /// # let store = Store::default();
+    /// #
+    /// let mt = MemoryType::new(1, None, false);
+    /// let m = Memory::new(&store, mt).unwrap();
+    ///
+    /// assert_eq!(m.ty(), mt);
+    /// ```
+    pub fn ty(&self) -> MemoryType {
+        let mut ty = self.vm_memory.ty.clone();
+        ty.minimum = self.size();
+        ty
+    }
+
+    /// Returns the [`Store`] where the `Memory` belongs.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # use wasmer_js::{Memory, MemoryType, Pages, Store, Type, Value};
+    /// # let store = Store::default();
+    /// #
+    /// let m = Memory::new(&store, MemoryType::new(1, None, false)).unwrap();
+    ///
+    /// assert_eq!(m.store(), &store);
+    /// ```
+    pub fn store(&self) -> &Store {
+        &self.store
+    }
+
+    /// Retrieve a slice of the memory contents.
+    ///
+    /// # Safety
+    ///
+    /// Until the returned slice is dropped, it is undefined behaviour to
+    /// modify the memory contents in any way including by calling a wasm
+    /// function that writes to the memory or by resizing the memory.
+    pub unsafe fn data_unchecked(&self) -> &[u8] {
+        unimplemented!("direct data pointer access is not possible in js");
```

Comment:
```suggestion
        unimplemented!("direct data pointer access is not possible in JavaScript");
```

I would suggest to basically remove the `data_unchecked` function, so that it emits a compiler error to the user, instead of a runtime error.